tests/test_assert.cpp

```diff
@@ -0,0 +1,18 @@
+#include "test_assert.h"
+
+int testing_fails = 0;
+
+void TestFail(const char *expval, const char *val, const char *exp,
```

Comment:
lets keep our "testing framework" header only if possible. since there is an `assert(0)` in there, it is intended to halt on the first test failure.

Reply:
It's not clear what the issue is with `assert(0)`. There are globals like `testing_fails`. I'm separating declaration and definition.  The `test.cpp` file is kind of too big.

Reply:
I meant, since it is intended to halt on first failure, there's no need to count fails, and no need for a global, thus this code can be kept in the header.

Reply:
I understand now. I prefer counting test failures rather than bailing out on the first error. This seems to be a norm in the gradle world (Java). But I could be persuaded otherwise.

I updated the TEST macros to include full function name as there are template instantiations. I tested without `assert(0)` and caused some tests to fail deliberately. I see output like this
```
/Users/sutambe/projects/myoss/myflatbuffers/cmake-build-debug/flattests
VALUE: "0"
EXPECTED: "1"
TEST FAILED: /Users/sutambe/projects/myoss/myflatbuffers/tests/test_builder.h:166, release_n_verify(b1, m2_name, m1_color) in static void BuilderTests<flatbuffers::FlatBufferBuilder>::builder_move_assign_after_releaseraw_test() [Builder = flatbuffers::FlatBufferBuilder]
VALUE: "0"
EXPECTED: "1"
TEST FAILED: /Users/sutambe/projects/myoss/myflatbuffers/tests/test_builder.h:166, release_n_verify(b1, m2_name, m1_color) in static void BuilderTests<TestHeapBuilder>::builder_move_assign_after_releaseraw_test() [Builder = TestHeapBuilder]
VALUE: "0"
EXPECTED: "1"
TEST FAILED: /Users/sutambe/projects/myoss/myflatbuffers/tests/test_builder.h:166, release_n_verify(b1, m2_name, m1_color) in static void BuilderTests<GrpcLikeMessageBuilder>::builder_move_assign_after_releaseraw_test() [Builder = GrpcLikeMessageBuilder]
3 FAILED TESTS

Process finished with exit code 1
```
Output on Windows is not gonna be so nice. It is still better as it tells me that 3 tests failed in all as opposed to the first one only. I'll rebase if you are ok with it or revert and drop the counter.

What say?